Implement -version parameter in console host (address part of https:/…

JamesWTruher opened the pull request:
#1084

This does not support providing a specific version to run, but
like most other *nix commands, -version will now return the version
of the PowerShell Engine. 'powershell' is prepended to the output to
match other *nix commands. We are using gitcommitid which includes more
info about the build.

JamesWTruher committed:
Implement -version parameter in console host (address part of https://github.com/PowerShell/PowerShell/issues/1084)

This does not support providing a specific version to run, but
like most other *nix commands, -version will now return the version
of the PowerShell Engine. 'powershell' is prepended to the output to
match other *nix commands. We are using gitcommitid which includes more
info about the build.

joeyaiello commented:
@PowerShell/powershell-committee is good with all of this. 👍

vors commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -551,6 +571,7 @@ private void ParseHelper(string[] args)
                 {
                     _sshServerMode = true;
                 }
+
```
redundant extra-line before `else`

vors commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -316,6 +316,14 @@ internal bool ServerMode
+        {
+            get
+            {
+                return _showVersion;
```
why not simple `{ get; private set; }`

daxian-dbw replied:
I think it's fine to follow the existing code pattern here. We can convert the style using a tool later.

JamesWTruher replied:
to match the current code style

---
In reply to: [100213197](https://github.com/PowerShell/PowerShell/pull/3115#discussion_r100213197) [](ancestors = 100213197)

vors commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -509,6 +516,19 @@ private void ParseHelper(string[] args)
                     switchKey = switchKey.Substring(1);
                 }
 
+                // If version is in the commandline, don't continue to look at any other parameters
```
I wonder how it's going to interact with existing in Windows PowerShell 
`powershell -version 2.0`

Since we are stopping to build Windows powershell, and core edition doesn't have any `-version 2.0` support, it's probably okey.

SteveL-MSFT replied:
@vors, the @PowerShell/powershell-committee discussed this and that is pretty much the thinking: `-version` for Windows PowerShell really only supported 2.0 and we're moving away from that.  Aligning with Linux convention makes sense.  PSCore supports side-by-side intrinsically and using file paths to launch different versions is completely acceptable.

joeyaiello replied:
Furthermore we can always do something like `powershell -version:2` if people really demand that functionality in the future (and actually @lzybkr and @BrucePay agreed that it would even be acceptable to do `powershell -version 2` as it's highly unlikely that anyone wants to echo one int when they shell out to PowerShell).

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -509,6 +516,19 @@ private void ParseHelper(string[] args)
+                    _noInteractive = true;
+                    _skipUserInit = true;
+                    _noExit = false;
+                    _commandLineCommand = "'powershell ' + $psversiontable.gitcommitid";
```
Why are you setting `_commandLineCommand` here?

JamesWTruher replied:
vestige of earlier implementation - removed

---
In reply to: [100401276](https://github.com/PowerShell/PowerShell/pull/3115#discussion_r100401276) [](ancestors = 100401276)

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHost.cs:
```diff
@@ -190,6 +190,15 @@ internal sealed partial class ConsoleHost
+                    // Alternatively, we could call s_theConsoleHost.UI.WriteLine(s_theConsoleHost.Version.ToString());
+                    // or start up the engine and retrieve the information via $psversiontable.GitCommitId
+                    // but this returns the semantic version and avoids executing a script
+                    s_theConsoleHost.UI.WriteLine("powershell " + PSVersionInfo.GetCommitInfo());
```
I recommend to use `PSVersionInfo.GetPSVersionTable()["GitCommitId"]`, or even better, add an internal static property `GitCommitId` right after the `PSVersion` property at [here](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/PSVersionInfo.cs#L148), and then use `PSVersionInfo.GitCommitId` here.

Reason: PSVersionInfo has a type initializer, which calls `GetCommitInfo()` and store the value in a hashtable. the current implementation of `GetCommitInfo()` reads git commit information from a file, which is slow and this implementation is subject to change (this method may be gone in near future -- we want to generate code at build time to embed the git commit id info, so as to avoid reading a file at startup time). Since the type initializer must run the first time you access a type, it's a waste to call `GetCommitInfo()` here to read the file again.

daxian-dbw commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -163,6 +163,20 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
+            $observed = & $powershell -version -command get-date
+            # no extraneous output
+            $observed | should be $currentVersion
+            
```
There is an extra blank line here.

daxian-dbw commented on test/powershell/Host/ConsoleHost.Tests.ps1:
```diff
@@ -163,6 +163,20 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
+
+        It "-Version should ignore other parameters" {
+            $currentVersion = "powershell " + $PSVersionTable.GitCommitId.ToString()
+            $observed = & $powershell -version -command get-date
```
what would happen if I have `powershell -noprofile -noexit -version`?

JamesWTruher replied:
it would report the version and exit - I can add a test for that explicitly if you like. Rather than testing all of the permutations for the parameters, I chose a representative single case

---
In reply to: [100405816](https://github.com/PowerShell/PowerShell/pull/3115#discussion_r100405816) [](ancestors = 100405816)

JamesWTruher committed:
use a new static internal for GitCommitId rather than calling method on PSVersionInfo

Remove extraneous blank lines in files
remove line which set the commandline when retrieving the version as it is unneeded

daxian-dbw reviewed (approved):
LGTM
